File: cocache-spring-redis/src/test/java/me/ahoo/cache/spring/redis/RedisDistributedCachingTest.java
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.cache.spring.redis;

import me.ahoo.cache.CacheValue;
import me.ahoo.cache.TtlAt;
import me.ahoo.cache.spring.redis.codec.StringToStringCodecExecutor;
import me.ahoo.cosid.util.MockIdGenerator;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.data.redis.connection.RedisStandaloneConfiguration;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import org.springframework.data.redis.core.StringRedisTemplate;

/**
 * RedisDistributedCachingTest .
 *
 * @author ahoo wang
 */
class RedisDistributedCachingTest {
    StringRedisTemplate stringRedisTemplate;
    StringToStringCodecExecutor codecExecutor;
    LettuceConnectionFactory lettuceConnectionFactory;
    RedisDistributedCache<String> distributedCaching;
    
    @BeforeEach
    private void setup() {
        RedisStandaloneConfiguration redisStandaloneConfiguration = new RedisStandaloneConfiguration();
        lettuceConnectionFactory = new LettuceConnectionFactory(redisStandaloneConfiguration);
        lettuceConnectionFactory.afterPropertiesSet();
        stringRedisTemplate = new StringRedisTemplate(lettuceConnectionFactory);
        codecExecutor = new StringToStringCodecExecutor(stringRedisTemplate);
        distributedCaching = new RedisDistributedCache<>(MockIdGenerator.INSTANCE.generateAsString(), stringRedisTemplate, codecExecutor);
    }
    
    @AfterEach
    void destroy() {
        if (null != lettuceConnectionFactory) {
            lettuceConnectionFactory.destroy();
        }
    }
    
    @Test
    void setThenGet() {
        final String CACHE_KEY = "setThenGet";
        final CacheValue<String> CACHE_VALUE = CacheValue.forever("setThenGet");
        distributedCaching.setCache(CACHE_KEY, CACHE_VALUE);
        Assertions.assertEquals(CACHE_VALUE, distributedCaching.getCache(CACHE_KEY));
    }
    
    @Test
    void getExpireAt() {
        final String CACHE_KEY = MockIdGenerator.INSTANCE.generateAsString();
        
        Long ttlAt = distributedCaching.getExpireAt(CACHE_KEY);
        Assertions.assertNull(ttlAt);
        final CacheValue<String> CACHE_VALUE = CacheValue.forever(CACHE_KEY);
        distributedCaching.setCache(CACHE_KEY, CACHE_VALUE);
        ttlAt = distributedCaching.getExpireAt(CACHE_KEY);
        Assertions.assertEquals(TtlAt.FOREVER, ttlAt);
    }

    @Test
    void setThenEvict() {
        final String CACHE_KEY = "setThenEvict";
        final CacheValue<String> CACHE_VALUE = CacheValue.forever("setThenEvict");
        distributedCaching.setCache(CACHE_KEY, CACHE_VALUE);
        distributedCaching.evict(CACHE_KEY);
        Assertions.assertNull(distributedCaching.get(CACHE_KEY));
    }
//
//    @SneakyThrows
//    @Test
//    void notifyInvalidate() {
//        final String CACHE_KEY = CacheKey.forever("notifyInvalidate");
//        final CacheValue<String> CACHE_VALUE = CacheValue.forever("notifyInvalidate");
//        CountDownLatch countDownLatch = new CountDownLatch(1);
//        RedisDistributedCaching<String> distributedCaching = new RedisDistributedCachingBuilder<String>()
//            .commandConnection(new RedisConnection<>(redisClient.connect()))
//            .codecExecutorProvider(StringToStringCodecExecutor::new)
//            .redirect(new RedisConnection<>(redisClient.connectPubSub()))
//            .build();
//        distributedCaching.get(CACHE_KEY);
//        distributedCaching.addListener(key -> {
//            if (CACHE_KEY.equals(key)) {
//                countDownLatch.countDown();
//            }
//        });
//        distributedCaching.set(CACHE_KEY, CACHE_VALUE);
//        Assertions.assertTrue(countDownLatch.await(2, TimeUnit.SECONDS));
//    }
//
//    @SneakyThrows
//    @Test
//    void notifyInvalidateWhenNoLoop() {
//        final String CACHE_KEY = CacheKey.forever("notifyInvalidateWhenNoLoop");
//        final CacheValue<String> CACHE_VALUE = CacheValue.forever("notifyInvalidateWhenNoLoop");
//        CountDownLatch countDownLatch = new CountDownLatch(1);
//        RedisDistributedCaching<String> distributedCaching = new RedisDistributedCachingBuilder<String>()
//            .commandConnection(new RedisConnection<>(redisClient.connect()))
//            .codecExecutorProvider(StringToStringCodecExecutor::new)
//            .redirect(new RedisConnection<>(redisClient.connectPubSub()))
//            .noloop()
//            .build();
//        distributedCaching.get(CACHE_KEY);
//        distributedCaching.addListener(key -> {
//            if (CACHE_KEY.equals(key)) {
//                countDownLatch.countDown();
//            }
//        });
//        distributedCaching.set(CACHE_KEY, CACHE_VALUE);
//        Assertions.assertFalse(countDownLatch.await(2, TimeUnit.SECONDS));
//    }
}

File: cocache-spring-redis/src/test/kotlin/me/ahoo/cache/spring/redis/RedisDistributedCachingTest.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.cache.spring.redis

import me.ahoo.cache.DefaultCacheValue
import me.ahoo.cache.distributed.DistributedCache
import me.ahoo.cache.spring.redis.codec.StringToStringCodecExecutor
import me.ahoo.cache.test.DistributedCacheSpec
import me.ahoo.test.asserts.assert
import org.assertj.core.data.Offset
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.data.redis.connection.RedisStandaloneConfiguration
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory
import org.springframework.data.redis.core.StringRedisTemplate
import java.util.*

/**
 * RedisDistributedCachingTest .
 *
 * @author ahoo wang
 */
internal class RedisDistributedCachingTest : DistributedCacheSpec<String>() {
    lateinit var stringRedisTemplate: StringRedisTemplate
    lateinit var codecExecutor: StringToStringCodecExecutor
    lateinit var lettuceConnectionFactory: LettuceConnectionFactory

    override fun createCache(): DistributedCache<String> {
        return RedisDistributedCache(
            stringRedisTemplate,
            codecExecutor,
        )
    }

    override fun createCacheEntry(): Pair<String, String> {
        return UUID.randomUUID().toString() to UUID.randomUUID().toString()
    }

    @BeforeEach
    override fun setup() {
        val redisStandaloneConfiguration = RedisStandaloneConfiguration()
        lettuceConnectionFactory = LettuceConnectionFactory(redisStandaloneConfiguration)
        lettuceConnectionFactory.afterPropertiesSet()
        stringRedisTemplate = StringRedisTemplate(lettuceConnectionFactory)
        stringRedisTemplate.afterPropertiesSet()
        codecExecutor = StringToStringCodecExecutor(stringRedisTemplate)
        super.setup()
    }

    @AfterEach
    fun destroy() {
        if (null != lettuceConnectionFactory) {
            lettuceConnectionFactory.destroy()
        }
    }

    /**
     * Redis 实现的 ttlAt 经 getExpire 重建，写读跨越秒边界时会有 ±1 秒漂移——
     * 覆盖 TCK 的精确相等断言为容差断言（与 CodecExecutorSpec 的既有做法一致）。
     */
    @Test
    override fun setWithTtl() {
        val (key, value) = createCacheEntry()
        cache[key].assert().isNull()
        val cacheValue = DefaultCacheValue.ttlAt(value, 5)
        cache.setCache(key, cacheValue)
        cache[key].assert().isEqualTo(value)
        cache.getTtlAt(key).assert().isCloseTo(cacheValue.ttlAt, Offset.offset(1))
    }

    @Test
    override fun setWithTtlAmplitude() {
        val (key, value) = createCacheEntry()
        cache[key].assert().isNull()
        val cacheValue = DefaultCacheValue.ttlAt(value, 5, 1)
        cache.setCache(key, cacheValue)
        cache[key].assert().isEqualTo(value)
        cache.getTtlAt(key).assert().isCloseTo(cacheValue.ttlAt, Offset.offset(1))
    }
}
